geopandas/array.py

```diff
@@ -1258,7 +1258,29 @@ def _values_for_argsort(self):
         ExtensionArray.argsort
         """
         # Note: this is used in `ExtensionArray.argsort`.
-        raise TypeError("geometries are not orderable")
+        from geopandas.tools.hilbert_curve import _hilbert_distance
+
+        if self.size == 0:
+            # TODO _hilbert_distance fails for empty array
+            return np.array([], dtype="uint32")
+
+        mask_missing = self.isna()
+        mask_empty = self.is_empty
+        mask = mask_missing | mask_empty
+        if mask.any():
+            geoms = self.copy()
+            indices = np.nonzero(~mask)[0]
+            if indices.size:
+                geom = self[indices[0]]
+            else:
+                # for all-empty/NA, just take random geometry
+                geom = shapely.geometry.Point(0, 0)
```

Comment:
I still have to add specific tests for the missing + empty (the base pandas ones only test for missing).

We probably also want to ensure that the missing geometries always sort at the end or start of the result, and so we should assign the empties the resulting min or max distance (-/+ some margin)

Reply:
What about using `pandas.UInt32Dtype` instead of `uint32` and `pandas.NA` for missing and empty values? That would solve the issue and ensure that they are consistently sorted at the end of the Series.

Reply:
The `_values_for_argsort` is a quite specific API, and so works a bit differently (which in the meantime has been clarified in the docstring in pandas: https://github.com/pandas-dev/pandas/blob/031e9bbc3a9db4ab78be8b477b15d7173a1f625e/pandas/core/arrays/base.py#L645-L651, I could copy that over to our docstring here (or a summary of it)): the missing values are handled on the pandas side (using `arr.isna()`, they will ensure those get masked out before sorting, and put at the end/beginning depending on the `na_position` keyword). So for the values being returned here, the values corresponding to missing values need to have some value, but doesn't matter which. 
The main problem is that for geometry dtype, we don't only have missing values, but also empty geometries as a special case. And those ideally also sort first/last just after/before the missing values. 

I could also override `argsort` here, so we have a bit more control over the exact behaviour. That would allow to let the empty geometries also follow the `na_position` keyword (if we want that, but maybe we could also say that empty geometries are always sorted first ("smallest")

Reply:
I went with the "empties are always sorted first ("smallest")), and added a bunch of comments. The code has become quite convoluted just to deal with the missing/empties ...